lib/mongoid/matcher/eq_impl.rb

```diff
@@ -23,10 +23,40 @@ module EqImpl
           end
 =end
         else
-          value == condition ||
-          value.is_a?(Array) && value.include?(condition)
+          # When doing a comparison with Time objects, compare using millisecond precision
+          if value.kind_of?(Time) && condition.kind_of?(Time)
+            time_eq?(value, condition)
+          elsif value.is_a?(Array) && condition.kind_of?(Time)
+            value.map do |v|
+              if v.kind_of?(Time)
+                time_rounded_to_millis(v)
+              else
+                v
+              end
+            end.include?(time_rounded_to_millis(condition))
+          else
+            value == condition ||
+            value.is_a?(Array) && value.include?(condition)
+          end
         end
       end
+
+      # Per https://docs.mongodb.com/ruby-driver/current/tutorials/bson-v4/#time-instances,
+      # > Times in BSON (and MongoDB) can only have millisecond precision. When Ruby Time instances
+      # are serialized to BSON or Extended JSON, the times are floored to the nearest millisecond.
+      #
+      # > Because of this flooring, applications are strongly recommended to perform all time
+      # calculations using integer math, as inexactness of floating point calculations may produce
+      # unexpected results.
+      #
+      # As such, perform a similar operation to what the bson-ruby gem does
+      module_function def time_eq?(time_a, time_b)
+        time_rounded_to_millis(time_a) == time_rounded_to_millis(time_b)
+      end
+
+      module_function def time_rounded_to_millis(time)
+        return time.utc._bson_to_i * 1000 + time.usec.divmod(1000).first
```

Comment:
@p-mongo I looked into it. It turned out that `_bson_to_i` is not defined for `ActiveSupport::TimeWithZone` class, and here `time.class == ActiveSupport::TimeWithZone`. In this case call to `_bson_to_i` would go through [ActiveSupport::TimeWithZone#wrap_with_time_zone](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/time_with_zone.rb#L582), retaining the time zone.

I see following action points here:
1. Fix `bson-ruby` gem by adding `_bson_to_i` method to `ActiveSupport::TimeWithZone` class. However, this will fix the problem only if Mongoid uses the latest bson gem.
2. Add  `_bson_to_i` method to `ActiveSupport::TimeWithZone` somewhere in Mongoid as a palliative until bson is fixed.
3. Or just keep the proposed workaround.

Since we already have some extensions for `ActiveSupport::TimeWithZone` in Mongoid, I think we can go with option 2.

Reply:
@jonhyman Could you please give me push permissions to braze-inc/mongoid repo?
If not possible, could you apply the following path to this PR - https://patch-diff.githubusercontent.com/raw/comandeo/mongoid/pull/2.patch

Reply:
@comandeo I applied the patch